fix!: ensure deleting message does not accidentally delete unrelated notifications

seanstrom opened the pull request:
fixes https://github.com/status-im/status-mobile/issues/17052
status-mobile PR: https://github.com/status-im/status-mobile/pull/21157
this PR is a follow up to: https://github.com/status-im/status-go/pull/5520/

## The Issue

* This PR attempts to resolve an issue with notifications being deleted for users after another user deletes a message inside a community.
  * The cause of the issue seems to be related to how notifications can reference a message with the `LastMessage` field. And if a message is deleted from a community and the notifications reference that message via `LastMessage`, then the notifications are also removed/deleted
    * We can see where the notifications are removed via `LastMessage` here: https://github.com/status-im/status-go/blob/bd0464d192971a284f29cf64c97a1914124307fe/protocol/activity_center_persistence.go#L51-L70
 
### Some Context
* It seems that `LastMessage` has been used for some notifications to represent the message for a `NewOneToOne` or `NewPrivateGroupChat` notification. The other activity-center notification types do not seem to use `LastMessage` field directly
* It's also important to note that the `LastMessage` field for notifications is not stored with the notification data. It is actually derived/joined from the `chat` data via the `chatID`. 
* Based on the git commit history, it seem the `LatestMessage` checks for deleting notifications were originally meant for detecting the `NewOneToOne` and `NewPrivateGroupChat` notifications. Here's the link to the original PR: https://github.com/status-im/status-go/pull/2854
* For context here's the main usage of the notification LastMessage field in status-mobile: https://github.com/status-im/status-mobile/blob/seanstrom/skip-audio-player-tests/src/status_im/contexts/shell/activity_center/notification/contact_requests/view.cljs#L68
* And here's the the main usage of the notification LastMessage field in status-go: https://github.com/status-im/status-desktop/blob/7c184f35002886947493070400500e7e05de74b6/src/app_service/service/activity_center/dto/notification.nim#L134-L138

### This PR Solution
* This PR tries to transition the activity notifications for `NewOneToOne` and `NewPrivateGroupChats` to use the `Message` field instead of the `LastMessage` field, which should remove the need to delete notifications via the `LatestMessage` field.
* This PR also attempts to focus on detecting `NewOneToOne` or `NewPrivateGroupChat` notification types that have references to a deleted message via `LastMessage`.
  * It might not be necessary to still detect these edge cases.

seanstrom committed:
fix!: ensure deleting message does not accidentally delete notifications that reference the message with last message

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (15)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_multiplication_x: | 02ae1c69 | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5789/1/) | 2024-08-30 14:27:27 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5789/1//consoleText) |
| :heavy_check_mark: | 02ae1c69 | [#1](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5789/1/) | 2024-08-30 14:28:51 | ~2 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5789/1//consoleText) |
| :heavy_check_mark: | 02ae1c69 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5789/1/) | 2024-08-30 14:29:58 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-240830-142600-02ae1c-pr5789.zip) |
| :heavy_check_mark: | 02ae1c69 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5789/1/) | 2024-08-30 14:30:55 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240830-142559-02ae1c-pr5789.zip) |
| :heavy_check_mark: | 02ae1c69 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5789/1/) | 2024-08-30 14:31:38 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240830-142558-02ae1c-pr5789.aar) |
| | | | | | | |
| :heavy_check_mark: | ba43762d | [#2](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5789/2/) | 2024-08-30 15:46:25 | ~1 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5789/2//consoleText) |
| :heavy_check_mark: | ba43762d | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5789/2/) | 2024-08-30 15:46:49 | ~1 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240830-154514-ba4376-pr5789.aar) |
| :heavy_check_mark: | ba43762d | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5789/2/) | 2024-08-30 15:48:23 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240830-154515-ba4376-pr5789.zip) |
| :heavy_check_mark: | ba43762d | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5789/2/) | 2024-08-30 15:49:03 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-240830-154519-ba4376-pr5789.zip) |
| :heavy_check_mark: | ba43762d | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5789/2/) | 2024-08-30 16:18:12 | ~32 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5789/2//consoleText) |
| | | | | | | |
| :heavy_multiplication_x: | b8a128ed | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5789/3/) | 2024-09-02 13:12:27 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5789/3//consoleText) |
| :heavy_check_mark: | b8a128ed | [#3](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5789/3/) | 2024-09-02 13:13:31 | ~2 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5789/3//consoleText) |
| :heavy_check_mark: | b8a128ed | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5789/3/) | 2024-09-02 13:13:38 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-240902-131119-b8a128-pr5789.zip) |
| :heavy_check_mark: | b8a128ed | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5789/3/) | 2024-09-02 13:13:39 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240902-131116-b8a128-pr5789.aar) |
| :heavy_check_mark: | b8a128ed | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5789/3/) | 2024-09-02 13:14:40 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240902-131118-b8a128-pr5789.zip) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | a86b0445 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5789/4/) | 2024-09-02 13:15:28 | ~1 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5789/4//consoleText) |
| :heavy_check_mark: | a86b0445 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5789/4/) | 2024-09-02 13:15:52 | ~1 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240902-131417-a86b04-pr5789.aar) |
| :heavy_check_mark: | a86b0445 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5789/4/) | 2024-09-02 13:16:13 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-240902-131422-a86b04-pr5789.zip) |
| :heavy_check_mark: | a86b0445 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5789/4/) | 2024-09-02 13:17:52 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240902-131456-a86b04-pr5789.zip) |
| :heavy_check_mark: | a86b0445 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5789/4/) | 2024-09-02 13:45:57 | ~31 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5789/4//consoleText) |
| | | | | | | |
| :heavy_check_mark: | d4716cdf | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5789/5/) | 2024-09-03 07:38:35 | ~1 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240903-073703-d4716c-pr5789.aar) |
| :heavy_check_mark: | d4716cdf | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5789/5/) | 2024-09-03 07:39:05 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-240903-073707-d4716c-pr5789.zip) |
| :heavy_check_mark: | d4716cdf | [#5](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5789/5/) | 2024-09-03 07:39:16 | ~2 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5789/5//consoleText) |
| :heavy_check_mark: | d4716cdf | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5789/5/) | 2024-09-03 07:44:15 | ~7 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240903-073704-d4716c-pr5789.zip) |
| :heavy_check_mark: | d4716cdf | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5789/5/) | 2024-09-03 08:09:13 | ~32 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5789/5//consoleText) |

seanstrom commented on protocol/activity_center_persistence.go:
```diff
@@ -61,7 +61,10 @@ func (db sqlitePersistence) DeleteActivityCenterNotificationForMessage(chatID st
 	for _, notification := range notifications {
 		if notification.LastMessage != nil && notification.LastMessage.ID == messageID {
-			withNotification(notification)
+			switch notification.Type {
+			case ActivityCenterNotificationTypeNewOneToOne, ActivityCenterNotificationTypeNewPrivateGroupChat:
+				withNotification(notification)
+			}
```
Here we're focusing on only removing notifications that might have a custom `LastMessage` from creating a `NewOneToOne` or `NewPrivateGroupChat` notification. Although it's likely that we won't need this anymore since we no longer use the `LastMessage` field when creating those notifications.

If it makes sense to delete this `if-block` then that's what I'll do next.

igor-sirotin replied:
Looks to me like `LastMessage` can be completely removed 👍

seanstrom commented on protocol/messenger_handler.go:
```diff
@@ -292,7 +292,7 @@ func (m *Messenger) createMessageNotification(chat *Chat, messageState *Received
 		ID:          types.FromHex(chat.ID),
 		Name:        chat.Name,
-		LastMessage: message,
+		Message:     message,
```
Here we're changing the creation of the `NewOneToOne` and `NewPrivateGroupChat` notifications to use the `Message` field instead of the `LastMessage` field.

seanstrom commented on protocol/activity_center_persistence_test.go:
```diff
@@ -228,6 +233,103 @@ func (s *ActivityCenterPersistenceTestSuite) Test_DeleteActivityCenterNotificati
 	s.Require().NoError(err)
 }
 
+func (s *ActivityCenterPersistenceTestSuite) Test_DeleteActivityCenterNotificationsForMessage_LastMessage() {
+	// docs: Create the temporary test-database that will be used to store chats, messages, and notifications
+	db, err := openTestDB()
+	s.Require().NoError(err)
```
Here we're creating another test case for specifically deleting NewOneToOne and NewPrivateGroupChat notifications via matching the `LastMessage` field with the deleted message ID. However, if we decide that we no longer need to support checking the LastMessage field when deleting notifications, then we can remove this test case.

seanstrom commented on protocol/activity_center_persistence_test.go:
```diff
@@ -205,17 +205,22 @@ func (s *ActivityCenterPersistenceTestSuite) Test_DeleteActivityCenterNotificati
-		notif, err = p.GetActivityCenterNotificationByID(id)
-		s.Require().NoError(err, notif.ID)
-		s.Require().True(notif.Deleted, notif.ID)
-		s.Require().True(notif.Dismissed, notif.ID)
-		s.Require().True(notif.Read, notif.ID)
-	}
+	notif, err = p.GetActivityCenterNotificationByID(nID2)
+	s.Require().NoError(err, notif.ID)
+	s.Require().True(notif.Deleted, notif.ID)
+	s.Require().True(notif.Dismissed, notif.ID)
+	s.Require().True(notif.Read, notif.ID)
+
+	// Check that notifications with LastMessage.ID == messages[1].ID will remain.
+	// Unless the notification is of type NewOneToOne or NewPrivateGroupChat.
+	notif, err = p.GetActivityCenterNotificationByID(nID3)
+	s.Require().NoError(err)
+	s.Require().False(notif.Deleted, notif.ID)
+	s.Require().False(notif.Dismissed, notif.ID)
+	s.Require().False(notif.Read, notif.ID)
```
Here we're updating the original test case for deleting a message to not expect that we should notifications when they reference the deleted message via the `LastMessage` field.

seanstrom commented on protocol/activity_center_persistence_test.go:
```diff
@@ -205,17 +205,22 @@ func (s *ActivityCenterPersistenceTestSuite) Test_DeleteActivityCenterNotificati
+	s.Require().True(notif.Read, notif.ID)
+
+	// Check that notifications with LastMessage.ID == messages[1].ID will remain.
+	// Unless the notification is of type NewOneToOne or NewPrivateGroupChat.
```
TODO: We'll potentially need to update this comment

seanstrom reviewed (commented):
Self Review 📖

igor-sirotin commented on protocol/activity_center_persistence_test.go:
```diff
@@ -205,17 +205,22 @@ func (s *ActivityCenterPersistenceTestSuite) Test_DeleteActivityCenterNotificati
-		s.Require().True(notif.Dismissed, notif.ID)
-		s.Require().True(notif.Read, notif.ID)
-	}
+	notif, err = p.GetActivityCenterNotificationByID(nID2)
```
Also add this to prevent potential `panic` on access to `nil` pointer:
```go
s.Require.NotNil(notif)
```

igor-sirotin replied:
In fact, when `err != ni` , `notif` will be `nil`.
So this line is not quite good:
```go
s.Require().NoError(err, notif.ID)
```

igor-sirotin reviewed (changes requested):
Thanks for the tests! ❤️ 
I'm happy with the changes, just a minor thing in tests about accessing `nil`.

seanstrom committed:
fix_: remove checks for NewOneToOne or NewPrivateGroupChat

seanstrom commented:
Awesome! thanks for the review @igor-sirotin 🙌 

I've updated the PR to remove the `LastMessage` checks, and I also updated the tests to guard against `nil` and avoid panics when inspecting the `err` value.

Let me know if these changes look good 🙏

ilmotta reviewed (approved):
LGTM!

Can we remove `LastMessage` entirely from the struct? The only places we set it now are when we unmarshal from the DB.

We would also need to create a migration to remove the column and update the mobile repo because we still have references to the last message. Does the desktop repo need adaptations? For these reasons it could be better to do the full clean up in a separate PR.

seanstrom commented:
> Can we remove `LastMessage` entirely from the struct? The only places we set it now are when we unmarshal from the DB.
>

Yeah I'm not sure why notifications would need to reference the last message. It seems that `LastMessage` is intended to always be the last message of the related chat (the latest when retrieved?).

But when we access `:last-message` on mobile, it's rarely accessed directly from the notification. It usually seems to be accessed directly from the `chat` data.

For context, I think these are three main uses of `:last-message` with a notification:
* https://github.com/status-im/status-mobile/blob/97d4edcf30a1b34f48a6bfae80488cf496638bea/src/status_im/contexts/shell/activity_center/notification/contact_requests/view.cljs#L65-L68
* https://github.com/status-im/status-mobile/blob/97d4edcf30a1b34f48a6bfae80488cf496638bea/src/status_im/contexts/shell/activity_center/notification/contact_requests/view.cljs#L108-L111
* https://github.com/status-im/status-mobile/blob/97d4edcf30a1b34f48a6bfae80488cf496638bea/src/status_im/contexts/shell/activity_center/notification/contact_requests/view.cljs#L166-L170

Fortunately, each of those usages will default to using the `:message` field before defaulting the `:last-message` value. So we don't have to make any changes to mobile yet, but we would definitely need to clean those up soon.
 
> We would also need to create a migration to remove the column and update the mobile repo because we still have references to the last message. Does the desktop repo need adaptations? For these reasons it could be better to do the full clean up in a separate PR.

So there doesn't seem to be a column for `last_message` in the notifications table, because that data is joined from the `chat` table I think. So if did remove the struct field, we would have to double-check the places mobile or desktop directly access the field from a notification.

That all said, I'm all for a separate clean up PR 🧹 😄

ilmotta commented:
> So there doesn't seem to be a column for `last_message` in the notifications table, because that data is joined from the `chat` table I think. So if did remove the struct field, we would have to double-check the places mobile or desktop directly access the field from a notification.

Indeed, the value is joined from the `chats` table 👍🏼

igor-sirotin reviewed (approved):
Thanks for the changes 👍

qfrank reviewed (approved):
Very detailed, thank you for the PR!

churik commented:
@igor-sirotin @seanstrom 
are those breaking changes for the desktop app? If so, what is the effect?
We need to make status-desktop PR with required changes prior to merging, like we do this for mobile.

I'm not completely sure about procedure, in our case we just create status-mobile PR, based on particular status-go commit.

cc @anastasiyaig

igor-sirotin commented:
@seanstrom @churik @anastasiyaig 

Yes, a status-desktop PR is required when doing a breaking change in status-go. 
This is the only way to make sure both clients are compatible with the change.
This is stated in the auto-comment above: https://github.com/status-im/status-go/pull/5789#issuecomment-2321433337

> are those breaking changes for the desktop app? If so, what is the effect?

I'm not sure what exactly is a breaking change here, but having `LastMessage` empty might break something indeed.

-----

I think the procedure should be:
1. PR in both clients MUST be opened, tested and approved before merging the status-go PR.
2. Communicate with desktop/mobile teams to make required changes 
(because desktop devs can't read clojure, and mobile devs can't read nim)

seanstrom commented:
Yes I think desktop would be affected by these changes, but I haven't verified if desktop had the same glitch that was reported in https://github.com/status-im/status-mobile/issues/17052

The changes for removing `LastMessage` could create some issues, though it seems desktop checks for the `Message` field before checking for the `LastMessage` field here: https://github.com/status-im/status-desktop/blob/7c184f35002886947493070400500e7e05de74b6/src/app_service/service/activity_center/dto/notification.nim#L134-L138

---

I merged this PR too quickly, and in the future I'll ping in the status-go channel for both QA teams to see a potentially breaking change.
